test/e2e/clusternetworkpolicy_test.go

```diff
@@ -0,0 +1,544 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
+	. "github.com/vmware-tanzu/antrea/test/e2e/utils"
+)
+
+// common for all tests.
+var p80 int
+var p81 int
+var allPods []Pod
+var k8s *Kubernetes
+var allTestList []*TestCase
+var pods, namespaces []string
+var podIPs map[string]string
+
+// provide enough time for policies to be enforced & deleted by the CNI plugin.
+const networkPolicyDelay = 2 * time.Second
+
+func failOnError(err error, t *testing.T) {
+	if err != nil {
+		log.Errorf("%+v", err)
+		k8s.Cleanup(namespaces)
+		t.Fatalf("test failed: %v", err)
+	}
+}
+
+type TestCase struct {
+	Name  string
+	Steps []*TestStep
+}
+
+type TestStep struct {
+	Name         string
+	Reachability *Reachability
+	CNPs         []*secv1alpha1.ClusterNetworkPolicy
+	Port         int
+	Duration     time.Duration
+}
+
+func initialize(t *testing.T, data *TestData) {
+	p80 = 80
+	p81 = 81
+	pods = []string{"a", "b", "c"}
+	namespaces = []string{"x", "y", "z"}
+
+	for _, podName := range pods {
+		for _, ns := range namespaces {
+			allPods = append(allPods, NewPod(ns, podName))
+		}
+	}
+	err := enableCNP(data)
+	failOnError(err, t)
+	k8s, err = NewKubernetes(data)
+	failOnError(err, t)
+	ips, err := k8s.Bootstrap(namespaces, pods)
+	failOnError(err, t)
+	podIPs = *ips
+}
+
+func enableCNP(data *TestData) error {
+	configMap, err := data.GetAntreaConfigMap(antreaNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get ConfigMap: %v", err)
+	}
+	antreaControllerConf, _ := configMap.Data["antrea-controller.conf"]
+	antreaAgentConf, _ := configMap.Data["antrea-agent.conf"]
+	antreaControllerConf = strings.Replace(antreaControllerConf, "#featureGates:", "featureGates:\n   ClusterNetworkPolicy: true", 1)
+	antreaAgentConf = strings.Replace(antreaAgentConf, "#featureGates:", "featureGates:\n   ClusterNetworkPolicy: true", 1)
+	configMap.Data["antrea-controller.conf"] = antreaControllerConf
+	configMap.Data["antrea-agent.conf"] = antreaAgentConf
+	if _, err := data.clientset.CoreV1().ConfigMaps(antreaNamespace).Update(context.TODO(), configMap, metav1.UpdateOptions{}); err != nil {
+		return fmt.Errorf("failed to update ConfigMap %s: %v", configMap.Name, err)
+	}
+	_, err = data.restartAntreaControllerPod(defaultTimeout)
+	if err != nil {
+		return fmt.Errorf("error when restarting antrea-controller Pod: %v", err)
+	}
+	// TODO: restart agent pods as well if featureGate affects agent
```

Comment:
@tnqn @jianjuns do you think we should enable CNPs by default in testing to avoid this ugliness, or does that compromise our test coverage for the "default" case (no CNP since it is an alpha feature).

My preference for the v0.8.0 time frame would be to define a new Github CI job for CNP testing on a Kind cluster. We would only run the CNP e2e tests as part of this job, and skip them for all the existing jobs. @Dyanngg what do you think?

Reply:
> this TODO?

Right now we do not need to restart agent.

Reply:
> @tnqn @jianjuns do you think we should enable CNPs by default in testing to avoid this ugliness, or does that compromise our test coverage for the "default" case (no CNP since it is an alpha feature).
> 
> My preference for the v0.8.0 time frame would be to define a new Github CI job for CNP testing on a Kind cluster. We would only run the CNP e2e tests as part of this job, and skip them for all the existing jobs. @Dyanngg what do you think?

I actually don't think it is bad to put CNP in the default e2e suite given that security_test.go also configures antrea controller as needed. If the e2e test cases turn out to be stable we could enable this for every PR, So that we can capture any breaking to the CNP feature at the PR  level.

Reply:
considering there is code sharing between CNP and K8s NP, it might still be a good idea to run CNPs e2e tests with incoming PRs.

Reply:
We should definitely run these tests for every PR. I'm just trying to avoid having to restart the agent on the master Node, it sometimes creates stability issues in Kind; I assume because br-phy gets destroyed (although these stability issues may all have been resolved or worked around by now, I am not sure).

I was suggesting doing it in a separate Github CI job. Currently (or at least after the proxy PR is merged), we have 6:
* encap 
* noEncap
* hybrid
* encap w Proxy
* noEncap w Proxy
* hybrid w Proxy

I was suggesting adding the following:
* encap w CNP

The first 6 jobs would not run the CNP test and the CNP feature would be turned off.
The 7th job would create a cluster and deploy Antrea with the CNP feature enabled, and the CNP tests would be run.

You can take a look at what was done for the Antrea proxy: https://github.com/vmware-tanzu/antrea/pull/772/files#diff-1f569612d6daadc573872784ea271152

Let me know what you think. I am also afraid that if we run all tests, including proxy & CNP in one job, it may start taking too long.

Reply:
@antoninbas I think it is the right direction to go. I will start working on adding a separate Github CI job. But in case we can't make it for this release, this e2e testsuite should be good as is, and people should be okay to cope with e2e tests that runs 5 more minutes for a couple of days :)

Reply:
We talked offline and decided to push it to after the release. See https://github.com/vmware-tanzu/antrea/issues/893